### DISCUSSION TITLE
Which step is correct sequence?

### DISCUSSION THREAD

**Discussion by lima4** [milvus-io/milvus#26106] [Q&A and General discussion]
Im testing milvus vector database in local.
I used on disk index to test continuously without a lot of memory capacity

Sequence 1

```
1. Create a collection with 3 varchar field and 1 embedding vector field(1536 dimensions)
2. Create index on the collection (DISKANN) before inserting any vectors
3. Insert the vectors
4. Index search with index_params
```

Sequence 2

```
1. Create a collection with 3 varchar field and 1 embedding vector field(1536 dimensions)
2. Insert the vectors
3. Create index on the collection (DISKANN, embedding vector) 
4. Index search with index_params
```

In the miluvs documentation, it seems to be written as sequence 2
But when I did sequence 2, an error occurred, so when I went to sequence 1, there was no problem

This is my error when i did sequence 2 (Create index on the embedding vector field)
<_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNAVAILABLE: ipv4:10.1.51.51:19530: Socket closed>

And docker-compose in docker-compose, milvus-standalone was shutting down

have any more docs or example about on-disk index?

**Comment by yhmo** [comment]
Milvus manages data by segments:
- By default, a segment size is around 512MB.
- For each segment, Milvus builds an independent index for it. Index is built asynchronously.
- After you call collection.load(). If a segment has no index, milvus will load the vectors into memory to do brute-force search. Else if a segment has been indexed, milvus will load the index into memory to do index search.
- When you call collection.search(), the search engine does ANN search for segments one by one. Search is in-memory.
- For each segment, the search engine generates topk results. Finally, all the topk results are merged into a final topk results and returned to the client.

No matter what kind of index you choose, there are 2 kinds of workflows for different use cases:
**A**:
1. Create a collection with 3 varchar field and 1 embedding vector field(1536 dimensions)
2. Create index on the collection (DISKANN) before inserting any vectors
3. Call collection.load() to tell milvus automatically load the data
4. Insert the vectors
5. Index search with index_params

**B**:
1. Create a collection with 3 varchar field and 1 embedding vector field(1536 dimensions)
2. Insert all the vectors
3. Create index on the collection (DISKANN, embedding vector) 
4. Call collection.load() to load the index into memory
5. Index search with index_params



**A** is for "dynamic data" ------- user insert rows continually, frequently, 10 rows per minute, etc. 
Call collection.create_index() at step2, it does nothing since no data at this point, just specify an index type for this collection.
Call collection.load() at step3, tells milvus auto load the increment data. Index is built asynchronously. When you call search(), some segments have not been indexed, then some segments are brute-force search, some segments are index search. If you want to get best search performance, you need to stop insertion and wait until all the index is done by `utility.wait_index_progress_done()`


**B** is for "static data" ------ the number of rows is a const value. the collection has only 1M vectors, etc.
Call collection.create_index() at step3, it will build index for all the segments since you have insert data at step2. collection.create_index() will return util all the index are finished.
Call collection.load() to load the index into memory at step4
Search performance is best since all the data is indexed.

**Reply by lima4** [reply]
thanks for the good explanation!!!!

By any chacne you know anything about this error?

```
<_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNAVAILABLE: ipv4:10.1.51.51:19530: Socket closed>

And docker-compose in docker-compose, milvus-standalone was shutting down
```

**Reply by yhmo** [reply]
Looks like the milvus instance is down. Maybe OOM. You can assign more memory resources to the container.
